topdown/walk.go

```diff
@@ -70,6 +80,31 @@ func walk(filter, path *ast.Array, input *ast.Term, iter func(*ast.Term) error)
 	return nil
 }
 
+func walkNoPath(path, input *ast.Term, iter func(*ast.Term) error) error {
+	if err := iter(ast.ArrayTerm(path, input)); err != nil {
+		return err
+	}
+
+	switch v := input.Value.(type) {
+	case ast.Object:
+		return v.Iter(func(_, v *ast.Term) error {
+			return walkNoPath(path, v, iter)
+		})
+	case *ast.Array:
+		for i := 0; i < v.Len(); i++ {
```

Comment:
Nit: `ast.Array` also has an `.Iter()` function. Right now, it doesn't do much more than what you do here, but using it would pick up any future optimizations, etc.
Unfortunately, we don't have an e.g. `ast.Iterable` interface that sets and arrays share.

Reply:
We could just create an ad-hoc interface and use that.

```go
type iterable interface {
	Iter(func(*ast.Term) error) error
}
```

But there would be more nits to pick in that solution than what's currently here 😄